Generalize sockets library and add test coverage for serviceLB.

tommyp1ckles opened the pull request:
The `datapath/sockets` code evolved from the use-case of serviceLB socket termination.  This makes the code very specific to that code. Instead this generalizes the underlying code to expose more general functions for iterating and closing sockets via netlink.


```release-note
Make datapath sockets code more general to promote code reuse.
```

tommyp1ckles committed:
sockets: use general purpose iterateNetlinkSockets function.

Instead of having a custom function for iterating UDP connections
rename and change this function to: iterateNetlinkSockets.

iterateNetlinkSockets will be able iterate any type of socket via
the netlink api.

In subsequent commits, this will be used to make the datapath/sockets
library a general purpose sockets library that allows for separately
iterating and destroying sockets via netlink.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
sockets: add test for Destroy function.

This is used to validate that subsequent changes do
not introduce regressions to functionality relying on
sockets.Destroy (i.e. svc-LB UDP sock destroy).

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
sockets: generalize destroySocket.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
sockets: add exported Iterate and DestroySocket functions.

Note: Up to this change was benchmarks to validate no memory
usage regressions are introduced following previous optimizations
in previous commits:

-------------
New:
-------------

goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/datapath/sockets
BenchmarkSocketReqSerialize-8           69173708               174.3 ns/op           128 B/op          2 allocs/op
BenchmarkSocketDeserialize-8            159207418               75.43 ns/op           64 B/op          4 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/datapath/sockets   31.865s

-------------
Old:
-------------

goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/datapath/sockets
BenchmarkSocketReqSerialize-8           69530899               172.2 ns/op           128 B/op          2 allocs/op
BenchmarkSocketDeserialize-8            163702172               73.25 ns/op           64 B/op          4 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/datapath/sockets   31.551s

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
mockmaps: add AddSockRevNat to mock lb map for upserting socket rev nat.

This will be used in subsequent commits for building tests for service
UDP socket termination functionality.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
netns: add function that returns pinned netns iterator.

This provides an easy iterator type for iterating netns in
the default pinned directory.
This being a simple function that returns the iterator will
allow for easy test mocking in subsequent commands where we
want to create real network-namespaces, but don't necessarily
need to pin them to the filesystem.

Errors are aggregated in a buffered channel so they can be
accessed during iteration.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
service: make service testable.

This abstracts concern related to iterating netns and
uses the new netns.All() function to replace the current
default functionality parity.

This will allow for providing an iterator that doesn't
rely on pinned netns (i.e. just ones we create at runtime
for the test process) which will make it easier to test.

As well, switch global use of *option.Config to a field
now provided via hive. This will avoiding global options
mutation for tests.

Finally, we now return errors from the terminate function,
although we do not handle it as errors are already logged.
The purpose is to return errors while testing in following
commits.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
service_test: add test to service lb connection handling.

This provides integration testing to test service manager
TerminateUDPConnectionsToBackend functionality together
with sockets.Destroy implementation.

Note: This is committed prior to further changes to
pkg/datapath/sockets code to validate that tests succeed
before and after the change.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

aditighag commented on pkg/netns/netns_linux.go:
```diff
@@ -243,3 +247,49 @@ func getFromPath(path string) (*os.File, error) {
+// namespaces should not be copied and used outside of the iterator loop.
+// The *NetNS handle is provided for to allow for cases where the ns must
+// be used explicitly*inside* the loop.
+func All() (iter.Seq2[string, *NetNS], <-chan error) {
```
If this is only used in testing, does it make sense to move this to `_test.go` file?

tommyp1ckles replied:
It's used for testing, but by the service package so it can't be in a _test file.

aditighag commented on pkg/datapath/sockets/sockets.go:
```diff
@@ -35,6 +35,46 @@ var (
+}
+
+// IterateAs iterates netlink sockets via a callback.
+func Iterate(proto uint8, family uint8, stateFilter uint32, fn func(*netlink.Socket, error) error) error {
```
>IterateAs, Iterate, DestroySocketAs, DestroySocket

I'm trying to understand if we need the other variants that deal with network namespaces. If I'm a new user of the sockets package, it's not obvious to me which variant I should use. 

>This choice was made to provide a more explicit way to manage network
namespaces.

`pkg/netns` provides APIs to manage and execute code inside network namespaces. I believe we can use the common API instead of creating other ways. What do you think?

tommyp1ckles replied:
Perhaps this deserves a seperate discussion, so I'm in favor of dropping these changes in this PR in order to not mix these proposed changes.

Specifically, what I'm worried about for using the `netns.Do` approach, say we have some code that is expected to execute inside the context of the network namespace:

```go
package ourPkg

import "github.com/cilium/cilium/pkg/somePkg"

ns.Do(func() {
  // ...
  somePkg.DoSomeStuff()
  // ...
})
```
 
The closure passed to Do executes with it's thread locked as the set network namespace call only applies to the specific process (thread). 

The concern is that lets say that the maintainer of somePkg makes some changes to somePkg, as a contrived example maybe they're trying to optimize some code to run in seperate goroutines for performance reasons:

before:

```go
func DoSomeStuff() {
  links, err := netlink.LinkList()
  if err != nil {
    panic(err) 
  }
  for _, link := range links {
    doSomethingWithLink(link)
  }
}
```

after:

```go
func DoSomeStuff() {
  links, err := netlink.LinkList()
  if err != nil {
    panic(err) 
  }
  wg := sync.WaitGroup{}
  for _, link := range links {
   wg.Add()
    go doSomethingWithLink(link, wg)
  }
  wg.Wait()
}
```

Unbeknownst to the maintainer of somePkg, they've now introduced a bug into `ourPkg` with a very reasonable change as the goroutines in `go doSomethingWithLink` may not be scheduled on the same thread as `DoSomeStuff()`.

Even worse, in practice this may not always happen in testing as new goroutines are often scheduled on the same thread as their parent).  This could introduce subtle bugs that may not show up reliably in testing. 

So the thought was that by creating families of functions that specify the netns, we can always be explicit with the network namespace we're using (having an implicit one using LockOSThread seems like not a good idea).

We can restrict this critical section to static bits of code we know won't suffer from this problem (i.e. not inside closures).  In the netlink library [this](https://github.com/cilium/cilium/pull/37600/files#diff-610ec9fa6c2a2ccf09f339e19271d1872768ae671bbf91a666d3af15e101d56fR302) should restrict this down to just the section sending the requests via the netlink socks.

Basically, this avoids having the situation where we're running code inside a context that can accidentally "break out" of the netns context.

tommyp1ckles replied:
But, this feels like a discussion that's worth having, otherwise this might just be introducing two ways of doing the same thing.

aditighag reviewed (changes requested):
It's great to see a generalized refactoring of the sockets package along with expanded test coverage. 
Initially, the socket-LB use case was the main driver for forcefully terminating sockets, but with new use cases emerging, consolidating the common logic makes a lot of sense. 
This PR takes that approach effectively, which is a solid step forward.

I've left some review comments on multiple exported functions in `pkg/sockets` based on my first pass.

aditighag commented on pkg/service/connections.go:
```diff
@@ -5,33 +5,31 @@ package service
 
-func (s *Service) TerminateUDPConnectionsToBackend(l3n4Addr *lb.L3n4Addr) {
+// TerminateUDPConnectionsToBackend closes UDP connection sockets that match a
+// l3/l4 tuple addr that also are tracked in the sock rev nat map (including netns cookie).
```
```suggestion
// TerminateUDPConnectionsToBackend closes UDP connection sockets that match the destination
// l3/l4 tuple addr that also are tracked in the sock rev nat map (including socket cookie).
```

We store socket cookies in the rev nat map, not netns cookies.

aditighag commented on pkg/service/service.go:
```diff
@@ -1774,6 +1782,9 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, isExtLocal, isIntLocal b
 			)
 		}
 		s.lbmap.DeleteBackendByID(id)
+		// Note: TerminateUDPConnectionsToBackend returns an error but we do not need to handle it here.
+		// errors are already logged inside the function and we do not want to return early in case of
+		// termination failures - these are always best effort.
```
Why return the error then? Is it mainly for the unit tests?

tommyp1ckles replied:
Yeah, we can refactor further but previously this would only log errors making it difficult to test.

tommyp1ckles replied:
Was in the interest of keeping functionality in parity as much as possible.

aditighag reviewed (commented):
Left a couple of minor comments 
Can you add sig-lb CODEOWNERS for `pkg/datapath/sockets/sockets.go`?

tommyp1ckles committed:
CODEOWNERS: add datapath/sockets to @sig-lb.

The only consumer of this code is (currently) service
manager so we should reflect that in reviews.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

aditighag commented on pkg/service/service_test.go:
```diff
@@ -2737,3 +2740,124 @@ func TestNotifyHealthCheckUpdatesSubscriber(t *testing.T) {
+	_, err = conn3.Read([]byte{0})
+	//nolint:errorlint
+	assert.True(t, err.(net.Error).Timeout(),
+		"other connection should not be closed due to mismatch on netns cookie")
```
*socket cookie?

aditighag commented on pkg/service/service_test.go:
```diff
@@ -2737,3 +2740,124 @@ func TestNotifyHealthCheckUpdatesSubscriber(t *testing.T) {
+	// ns3 will match revnat, but with a different cookie value
+	// so we expect to avoid a socket close (i.e. this is
+	// the case where we have matching tuple values, but
+	// in an unexpected network namespace).
```
" so we expect to avoid a socket close (i.e. this is
	// the case where we have matching tuple values, but
	// in an unexpected network namespace)."

Not sure what you mean by "an unexpected network namespace. As mentioned, we match socket cookie and not netns cookie. 

I would suggest making slight changes in order to make it a realistic unit test case - add connections in host netns and pod netns and check whether the expected connections are closed based on the config `option.Config.BPFSocketLBHostnsOnly`.

aditighag reviewed (approved):
Solid step in generalizing the sockets termination logic. 🚀 

It'll be good to fix the inconsistencies in the unit tests.
